crates/uv-requirements/src/discovery.rs

```diff
@@ -0,0 +1,579 @@
+use std::collections::BTreeMap;
+use std::path::{Path, PathBuf};
+
+use glob::{glob, GlobError, PatternError};
+use tracing::{debug, trace};
+
+use uv_fs::Simplified;
+use uv_normalize::PackageName;
+use uv_warnings::warn_user;
+
+use crate::pyproject::{PyProjectToml, Source, ToolUvWorkspace};
+use crate::RequirementsSource;
+
+#[derive(thiserror::Error, Debug)]
+pub enum DiscoverError {
+    #[error("No `pyproject.toml` found in current directory or any parent directory")]
+    MissingPyprojectToml,
+
+    #[error("Failed to find directories for glob: `{0}`")]
+    Pattern(String, #[source] PatternError),
+
+    #[error("Invalid glob: `{0}`")]
+    Glob(String, #[source] GlobError),
+
+    #[error(transparent)]
+    Io(#[from] std::io::Error),
+
+    #[error(transparent)]
+    Toml(#[from] toml::de::Error),
+
+    #[error("No `project` section found in: `{}`", _0.simplified_display())]
+    MissingProject(PathBuf),
+}
+
+/// A package in a workspace.
+#[derive(Debug, Clone)]
+#[cfg_attr(test, derive(serde::Serialize))]
+#[allow(dead_code)] // TODO(konsti): Resolve workspace package declarations.
+pub struct WorkspaceMember {
+    /// The path to the project root.
+    root: PathBuf,
+    pyproject_toml: PyProjectToml,
+}
+
+/// A package and the workspace it is part of.
+#[derive(Debug, Clone)]
+#[cfg_attr(test, derive(serde::Serialize))]
+#[allow(dead_code)] // TODO(konsti): Resolve workspace package declarations.
+pub struct ProjectWorkspace {
+    /// The path to the project root.
+    project_root: PathBuf,
+    /// The name of the package.
+    project_name: PackageName,
+    /// The path to the workspace root.
+    workspace_root: PathBuf,
+    /// The members of the workspace.
+    workspace_packages: BTreeMap<PackageName, WorkspaceMember>,
+    /// The source table for the workspace declaration.
+    workspace_sources: BTreeMap<PackageName, Source>,
+}
+
+impl ProjectWorkspace {
+    /// Find the current project and workspace.
+    pub fn discover(path: impl AsRef<Path>) -> Result<Self, DiscoverError> {
+        let Some(project_root) = path
+            .as_ref()
+            .ancestors()
+            .find(|path| path.join("pyproject.toml").exists())
+        else {
+            return Err(DiscoverError::MissingPyprojectToml);
+        };
+
+        debug!("Project root: `{}`", project_root.simplified_display());
```

Comment:
Can you make this message a bit more user-facing e.g. "Found project root `{}`" or similar?